packages/ens-utils/src/registrar.ts

```diff
@@ -0,0 +1,217 @@
+   */
+  validity?: TimePeriod;
+}
+
+export interface RegistrarBaseFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.BaseFee;
+  price: Price;
+  reason?: undefined;
+  validity?: undefined;
+}
+
+export interface RegistrarSpecialNameFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.SpecialNameFee;
+  price: Price;
+  reason: string;
+  validity?: undefined;
+}
+
+export interface RegistrarTemporaryFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.TemporaryFee;
+  price: Price;
+  reason: string;
+  validity: TimePeriod;
+}
+
+export interface RegistrarServiceFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.ServiceFee;
+  price: Price;
+  reason?: string;
+  validity?: undefined;
+}
+
+export type RegistrarCharge = RegistrarBaseFee | RegistrarSpecialNameFee | RegistrarTemporaryFee | RegistrarServiceFee;
+
+/**
+ * A price quote from a `Registrar` to register or renew a domain for a given
+ * period of time.
+ */
+export interface AbstractRegistrarPriceQuote {
+
+  /**
+   * The action associated with this `RegistrarPriceQuote`.
+   */
+  action: RegistrarAction;
+
+  /**
+   * The `TimePeriod` that this `RegistrarPriceQuote` is for.
+   */
+  rentalPeriod: TimePeriod;
+
+  /**
+   * The set of distinct `RegistrarCharge` values that must be paid to the
+   * `Registrar` to rent the domain for the period of `rentalPeriod`.
+   * 
+   * May be empty if the domain is free to rent for `rentalPeriod`.
+   */
+  charges: RegistrarCharge[];
+
+  /**
+   * The total price to rent the domain for the period of `rentalPeriod`.
+   * 
+   * This is the sum of `charges` in the unit of currency that should be paid
+   * to the registrar.
+   */
+  totalPrice: Price;
+}
+
+export interface RegistrationPriceQuote extends AbstractRegistrarPriceQuote {
+  action: typeof RegistrarAction.Register;
+}
+
+export interface RenewalPriceQuote extends AbstractRegistrarPriceQuote {
+  action: typeof RegistrarAction.Renew;
+}
+
+export interface Registrar {
+  getManagedSubname(name: ENSName): ENSName | null;
+
+  // NOTE: Throws RegistrarUnsupportedNameError if name is not supported by this registrar
+  getValidatedSubname(name: ENSName): ENSName;
+
+  getOnchainRegistrar(): ContractRef | null;
+
+  canRegister(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration?: Registration): boolean;
+  getRegistrationPriceQuote(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration?: Registration): RegistrationPriceQuote;
+  
+  canRenew(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration: Registration): boolean;
+  getRenewalPriceQuote(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration: Registration): RenewalPriceQuote;
+}
+
+export interface OnchainRegistrar extends Registrar {
+  getOnchainRegistrar(): ContractRef;
+}
+
+export interface OffchainRegistrar extends Registrar {
```

Comment:
Might be good to document this is a future-support-want feature that is not ready to be used?

Reply:
Sure you're welcome to do that.